### PR TITLE
Fix bugs in new task for reindexing AIPs from Storage Service

### PR DESCRIPTION
**Pull request by tw4l**
Fixes two bugs in the new AIP reindexing command added in commit 443fc2e66a101a0160951f06fa2f3a01717c9ce0:

* The task was not previously filtering out replicated AIPs, which was leading to a `lxml.etree.XMLSyntaxError` error for replicated AIPs during indexing (specifically, in the `index_aips_and_files` function in `elasticSearchFunctions.py`)
* The `--uuid` flag to reindex a single AIP was throwing an `AttributeError` because it was trying to get a property from a list, rather than from the first member of that list, and was incorrectly expecting `process_package` to return a tuple rather than a boolean.

Bugs addressed in this PR:
* https://github.com/archivematica/Issues/issues/1262
* https://github.com/archivematica/Issues/issues/1263

Connected to https://github.com/archivematica/issues/issues/734

### CODE REVIEW

File: src/archivematicaCommon/lib/archivematicaFunctions.py
Comment (ross-spencer) [inline src/archivematicaCommon/lib/archivematicaFunctions.py]: This might be a useful helper in `AMClient` where we already have a helper for compressed packages as an example. I am not convinced it sits particularly well here as its own helper in Archivematica given it is just a single list comprehension. 

As this is dealing with storage service functions, a nice approach might be to modify the storage service helper. 

* `storageService.get_file_info(package_type="AIP")`

If you were interacting with an API, you might see a filter. Something like:

* `storageService.get_file_info(package_type="AIP", filter_replicas=True)`
* `storageService.get_file_info(package_type="AIP", filter=REPLICAS)` (where replicas is an `enum` like value.

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (ross-spencer) [inline src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: Is this `black` formatted?

Comment (ross-spencer) [reply src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: An `AttributeError` is mentioned in the PR comment, is this the error that we're covering here? Or is this a tangential issue?

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (ross-spencer) [inline src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: As the return from this function will be `None`. We should consider removing the explicit return here.

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (ross-spencer) [inline src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: We can remove this `return` too.

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (ross-spencer) [inline src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: Shall we continue to test for `True` here like in the original line at `148`? If it isn't true, can we return immediately after 147? Or do we still need to commit to clean up? Is there any more information that the user (or us) need here, will we get enough information to track back from: `Failure. Indexing complete.` do you think?

**Review by ross-spencer — COMMENTED** [review]
Hi @tw4l I've left a few comments which I wouldn't mind getting your feedback on. 

In terms of process, I'm surprised this PR hasn't got an issue associated with it. Or, to be clear, I think archivematica/Issues#734 might otherwise have been moved into `review` and then one new issues associated with the new changes to this work so that there was good visibility prior to this PR, and simply because they are new issues. 

I think we've done it this way previously (without an issue) for your previous PR, but it can lead to rushing out the work we're doing where we should measure twice and cut once as much as possible. 

I'll drop you a line in Slack too.

File: src/archivematicaCommon/lib/archivematicaFunctions.py
Comment (ross-spencer) [reply src/archivematicaCommon/lib/archivematicaFunctions.py]: It dawns on me too, is there a way to only filter once? What information are we getting from the `storage_service.get_file_info()`? 

Is it possible to:

1. Retrieve all AIPs and AICs.
2. Filter replicated packages.
3. Direct the workflow in these functions by testing here whether something is an AIC or an AIP? 

It should simplify the code in rebuild AIP index, and then mean only filtering once. Basically we can remove some complexity (possibly).

Comment (tw4l) [reply src/archivematicaCommon/lib/archivematicaFunctions.py]: Hi @ross-spencer - thanks, this is good insight. I've taken the spirit of that you've said here and tried to address it in commit 7cca2c9624865044c95e160104a729fe01777361, which makes a single call to the storage service for packages, does all of the filtering in one shot (and moves that helper to `storageService.py`, where I think it's a better fit), and manages to remove a fair bit of duplicative code in the process. Let me know what you think.

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (tw4l) [reply src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: The `AttributeError` was being raised during `package_type = package_info.get("package_type")`, because `package_info` was the list returned by the Storage Service.

In this PR, I changed it so that `package_info` is the first item of the returned list: `package_info = storageService.get_file_info(uuid=uuid)[0]`. If there are no items in the list (i.e. a package with the specified UUID was not found in the Storage Service), it will raise an `IndexError`, so that's what I'm catching here.

I added some more information on the original bugs in this issue: https://github.com/archivematica/Issues/issues/1263

Comment (tw4l) [reply src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: The multi-line comment was run through `black`, but I had split the lines to try to keep line size reasonable. The changes made since have made that moot.

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (tw4l) [reply src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: The return will be `None`, but we need this return so that the task only indexes the one package we requested with the `--uuid` tag. Otherwise, the task will index this one package, and then move on to indexing all AIPs.

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (tw4l) [reply src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: See above.

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (tw4l) [reply src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: I think this is cleaned up a bit now. The messages here were a bit vague because `process_package` outputs more detailed information for each package to the terminal as well.

**Comment by tw4l** [comment]
Thank you @ross-spencer! This is ready for re-review.

I've created issues for each of the bugs being addressed in this PR and linked them in the PR description up top. Moving forward I will stick to the process of filing an issue first, then the related PR. I take your point about measuring twice and cutting once.

Some details on the changes made here are in the in-line comments above, but in general I have refactored to:
* Handle truthiness consistently (with `if x` instead of `if x is True`, especially where we absolutely know values will be booleans)
* Move the logic for indexing a single package to its own function
* Retrieve package info from the Storage Service and filter it once (with a modified version of the previous filtering helper, which I've now moved from `archivematicaFunctions.py` to `storageService.py`, which seems like a better home for it. This also required a slight tweak to the transfer reindexing task). Ideally this filtering could all happen as part of the call to the Storage Service but that would require more major changes, and I hope you will agree that this is a good compromise. It will be a bit faster and, as you suggested, simplifies the `handle` function, which should pay off in terms of readability and maintainability.

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (tw4l) [reply src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: Ordering went off with the refactor, so that comment might not be so helpful. I believe this return is necessary to prevent the task from continuing to index other packages when the --uuid flag is used.

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (ross-spencer) [reply src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: Cool. It's just a little jarring as we have `Flake8` [ignoring line-lengths](https://github.com/artefactual/archivematica/blob/396015a44723e6b2dd2f4d7bed1443514a429c12/tox.ini#L38) (E501) up to a certain point to be [compatible with `black`](https://stackoverflow.com/a/59184533) so it's a bit more obvious when the line hasn't been auto-formatted.

File: src/archivematicaCommon/lib/storageService.py
Comment (ross-spencer) [inline src/archivematicaCommon/lib/storageService.py]: What do you think about removing these five lines and making them the default parameters in the function signature?
```python
def filter_packages(
    package_list,
    statuses=("UPLOADED", "DEL_REQ"),
    package_types=("AIP", "AIC", "transfer", "DIP"),
    pipeline_uuid=None,
    filter_replicas=False,
): 
```
It seems like the logic would be about the same? It would really only have an impact if someone supplied `None` to the function, which might be a bit odd, but IDK?

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (ross-spencer) [reply src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: That makes sense. I feel like when there's a return like this in the function it says there might be something not quite right with the flow-control. So, in the code above we're directing the flow if `uuid` is set. Given that, I'd almost be tempted to say:
```python
if options["uuid"]:
   ...do thing
else:
   ...do the other thing
```
But that doesn't quite feel right either. I suppose a `return` does keep everything to the left quite nicely. 

But then I thought to myself what else might we be missing.

I had one observation, and that was the setup for Elasticsearch came too early. What happens if we retrieve zero packages from the storage service? e.g. there's a connection error? I think we will have deleted our index and didn't need to.

The second observation I had with the new function for indexing a single package is that we really wanted to do the same thing across both main sections of the code. Given the storage service helper always returns a list, we can treat a list of 1 (or zero) the same we treat a list of 100, with a few small exceptions written along the way.

I have modified your functions here. So I have removed `index_single_package` and moved the bulk of the unique work inline so that most of the logic is the same for bulk or single operations. Also, we now only setup Elasticsearch just in time, so that we only do anything with the index if there is something to do. See what you think:

```diff
diff --git a/src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py b/src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
index f766ee77..d48c05fa 100644
--- a/src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
+++ b/src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
@@ -116,9 +116,6 @@ class Command(DashboardCommand):
         logging.getLogger("elasticsearch").setLevel(logging.ERROR)
         logging.getLogger("archivematica.common").setLevel(logging.ERROR)
 
-        # Setup es_client and delete indices if required.
-        es_client = setup_es_for_aip_reindexing(self, options["delete_all"])
-
         # Create temporary directory for downloaded METS files.
         temp_dir = tempfile.mkdtemp()
 
@@ -129,29 +126,27 @@ class Command(DashboardCommand):
         if options["delete"]:
             delete_before_reindexing = True
 
-        # Index a single package and return if uuid option is enabled.
-        if options["uuid"]:
-            index_success = self.index_single_package(
-                options["uuid"], es_client, temp_dir, delete_before_reindexing
-            )
-            if index_success:
-                self.success("Success! Indexing complete.")
-            return
-
         self.info("Rebuilding 'aips' and 'aipfiles' indices")
-
         packages_not_indexed = []
-
-        # Index AIPs and AICs that are associated with this pipeline
-        # that are not deleted or replicas.
-        packages = storageService.get_file_info()
-        aips_to_index = storageService.filter_packages(
-            packages,
-            package_types=PACKAGE_TYPES_TO_INDEX,
-            pipeline_uuid=pipeline_uuid,
-            filter_replicas=True,
-        )
+        if not options["uuid"]:
+            # Index AIPs and AICs that are associated with this pipeline
+            # that are not deleted or replicas.
+            packages = storageService.get_file_info()
+            aips_to_index = storageService.filter_packages(
+                packages,
+                package_types=PACKAGE_TYPES_TO_INDEX,
+                pipeline_uuid=pipeline_uuid,
+                filter_replicas=True,
+            )
+        else:
+            aips_to_index = storageService.get_file_info(uuid=options["uuid"])
         aips_to_index_count = len(aips_to_index)
+        if aips_to_index_count > 1:
+            # For bulk operations setup es_client and delete indices if required.
+            es_client = setup_es_for_aip_reindexing(self, options["delete_all"])
+        elif aips_to_index_count:
+            # For single item operations, setup es_client and don't delete indices.
+            es_client = setup_es_for_aip_reindexing(self, False)
         aip_indexed_count = 0
         for aip in aips_to_index:
             is_aic = False
@@ -184,37 +179,6 @@ class Command(DashboardCommand):
                 )
             )
 
-    def index_single_package(self, uuid, es_client, temp_dir, delete_before_reindexing):
-        """Index single package by UUID.
-
-        :param uuid: Package UUID.
-        :param es_client: Elasticsearch client.
-        :param temp_dir: Path to tempdir for downloaded METS files.
-        :param delete_before_reindexing: Boolean of whether to delete
-        package from indices prior to reindexing.
-
-        :returns: Boolean indicating success.
-        """
-        self.info("Reindexing package {}".format(uuid))
-        try:
-            package_info = storageService.get_file_info(uuid=uuid)[0]
-        except IndexError:
-            self.error("Error indexing package {0}. UUID not found.".format(uuid))
-            return False
-
-        is_aic = False
-        package_type = package_info.get("package_type")
-        if package_type and package_type == "AIC":
-            is_aic = True
-
-        index_success = self.process_package(
-            es_client, package_info, temp_dir, delete_before_reindexing, is_aic=is_aic
-        )
-
-        self.info("Cleaning up")
-        shutil.rmtree(temp_dir)
-        return index_success
-
     def process_package(
         self, es_client, package_info, temp_dir, delete_before_reindexing, is_aic=False
     ):
```
You might see other optimizations, but does that help at all? (I wasn't quite sure how to write this down in a better narrative, so hoped maybe the code-diff would work as well).

**Review by ross-spencer — COMMENTED** [review]
Thanks @tw4l. It's all looking a lot cleaner. Nice work bringing those two functions together for the storage service helpers. 

There's a chunky comment/demonstration inline, let me know if that helps. And I hope it's not overreaching too much. I was wrestling with how to jot the point down more effectively.

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (tw4l) [reply src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: Yeah, fair!

File: src/archivematicaCommon/lib/storageService.py
Comment (tw4l) [reply src/archivematicaCommon/lib/storageService.py]: Yeah, agreed that making these the default parameters makes it cleaner and easier to read, and from the docstring one should know not to supply `None` so I think that's an okay tradeoff.

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (tw4l) [reply src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: It did help! I refactored based on your comments here (FWIW, the diff really did help me come around to your point) and the resulting code is shorter, more readable, and handles the case of no packages being returned by the Storage Service much more gracefully.

**Comment by tw4l** [comment]
Hi @ross-spencer - thanks, the inline diff really did help and your thinking brought some really nice changes to the code (I also added you as a co-author on the last commit, and will for the merge commit as well if that's okay with you).

I've flagged this for re-review - let me know if you'd like to see other changes.

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (ross-spencer) [inline src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py:130]: My instinct is still to protect against deletion of the entire index, if the user is only re-indexing a single UUID. The impact of that mistake goes up quite high the bigger the index is. Right now, if `--delete-all` is selected, also `-u` that's definitely possible in this script. What do you think about avoiding that?

There are various ways to do this, but maybe setting `options["'delete_all"]` to `False` here would be enough?

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (ross-spencer) [inline src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]:
```suggestion
                "Indexing complete. Indexed {} of {} AIPs/AICs. Packages not indexed: {}.".format(
```
I think you want `{2}` here! :wink: that being said it might be a good (inadvertent) demonstration of where we can let the positional arguments do the work, and avoid explicit indexes where possible.

I'd chose `{}` in this case. For other instances if you're reusing variables during string concatenation, then the named approach works quite nicely as a self-describing and very explicit option:
```python
            self.error(
                "Indexing complete. Indexed {demo_1} of {demo_2} AIPs/AICs. Packages not indexed: {demo_3}.".format(
                    demo_1=aip_indexed_count,
                    demo_2=aips_to_index_count,
                    demo_3=", ".join(packages_not_indexed),
                )
            )
```
There's a whole bunch of string concatenation methods but I tend to stick with this one for AM and personally.

**Review by ross-spencer — APPROVED** [review]
Nearly there @tw4l with one definite change needed for a string concat further below, and a request to change the logic around deletion a little bit. Let me know if you have any queries. 

Also, I had forgotten to do this, maybe you were already doing it? But `pylint --py3k`
```
$ pylint --py3k src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Using config file /home/ross-spencer/.pylintrc

------------------------------------
Your code has been rated at 10.00/10
``` 
perfect score: :100: /:100: !!

Approving so you can fix those up and get onwards and upwards once those changes are in.

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (tw4l) [reply src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py:130]: Honestly I went back and forth on this because I didn't want to unnecessarily limit users, but ultimately I think you're right that the impact of that mistake outweighs the remote possibility that someone would want to delete their entire index and reindex only one package (especially since I can't think of a use case for the latter). I'll make the change before merging.

File: src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py
Comment (tw4l) [reply src/dashboard/src/main/management/commands/rebuild_aip_index_from_storage_service.py]: Good catch! This named option is nice. It reminds me of f-strings, which I tend to prefer when possible.

**Pull request merged**
